language/move-vm/types/src/values/values_impl.rs

```diff
@@ -670,7 +677,30 @@ impl ContainerRef {
                 }
 
                 match self.container() {
-                    Container::Struct(r) => assign!(r, Struct),
+                    Container::Struct(r, ref_tag) => {
+                        let r1 = match c {
+                            Container::Struct(v, tag) => {
+                                if *ref_tag != tag {
```

Comment:
> This would be a potentially better/alternative/additional mitigation to recursing through the entire vector in `check_type`

I just added the check! It was surprisingly a one line change. I'll see how I can potentially relax the check_type